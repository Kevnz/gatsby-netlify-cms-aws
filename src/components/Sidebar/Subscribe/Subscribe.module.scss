@import '../../../assets/scss/variables';
@import '../../../assets/scss/mixins';

.subscribe {
  @include margin-bottom(1);
}
