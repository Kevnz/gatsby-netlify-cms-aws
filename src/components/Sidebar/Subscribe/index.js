// @flow
export { default } from './Subscribe';
