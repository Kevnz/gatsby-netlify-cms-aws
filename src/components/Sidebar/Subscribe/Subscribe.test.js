// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup } from '@testing-library/react';
import Subscribe from './Subscribe';

const setup = () => {
  const utils = render(<Subscribe />);
  const input = utils.container.querySelector('#email');
  const form = utils.container.querySelector('#mailchimpForm');
  return {
    input,
    form,
    ...utils,
  };
};

jest.mock('jsonp');
jest.useFakeTimers();
// jest.mock('semantic-ui-react/dist/es/modules/Transition/Transition.js');

afterEach(cleanup);

describe('Subscribe', () => {
  const endpoint =
    'https://domain.us3.list-manage.com/subscribe/post?u=aaaaaaa&amp;id=bbbbb';

  beforeEach(() => {
    process.env.GATSBY_MAILCHIMP_ENDPOINT = endpoint;
    jest.clearAllMocks();
  });

  it('renders correctly no endpoint', () => {
    process.env.GATSBY_MAILCHIMP_ENDPOINT = '';
    const tree = renderer.create(<Subscribe />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with endpoint', () => {
    const tree = renderer.create(<Subscribe />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Should update email with value', () => {
    const { input, container } = setup();
    fireEvent.change(input, { target: { value: 'admin@test.com' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Should subscribe to email', () => {
    const jsonp = require('jsonp');

    jsonp.mockImplementation((url, params, callback) => {
      callback(null, { status: 'success', message: 'message' });
    });

    const { input, form, container } = setup();
    const email = 'admin@test.com';
    fireEvent.change(input, { target: { value: email } });
    fireEvent.submit(form);

    expect(jsonp).toHaveBeenCalledTimes(1);
    expect(jsonp).toHaveBeenCalledWith(
      endpoint.replace('/post?', '/post-json?') +
        `&EMAIL=${encodeURIComponent(email)}`,
      { param: 'c', timeout: 3500 },
      expect.any(Function),
    );

    expect(container.firstChild).toMatchSnapshot();
    jest.runTimersToTime(4000);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Should handle jsonp error', () => {
    const jsonp = require('jsonp');

    jsonp.mockImplementation((url, params, callback) => {
      callback('Unknown error', null);
    });

    const { input, form, container } = setup();
    const email = 'admin@test.com';
    fireEvent.change(input, { target: { value: email } });
    fireEvent.submit(form);

    expect(jsonp).toHaveBeenCalledTimes(1);

    expect(container.firstChild).toMatchSnapshot();
    jest.runTimersToTime(4000);
    expect(container.firstChild).toMatchSnapshot();
  });
});
