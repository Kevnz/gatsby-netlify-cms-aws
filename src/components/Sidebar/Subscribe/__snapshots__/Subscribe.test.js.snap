// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribe Should handle jsonp error 1`] = `
<div
  class="subscribe"
>
  <form
    class="ui mini loading form"
    id="mailchimpForm"
  >
    <div
      class="unstackable fields"
    >
      <div
        class="required ten wide field"
      >
        <div
          class="ui input"
        >
          <input
            aria-label="email"
            id="email"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="admin@test.com"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <button
          class="ui mini secondary button"
        >
          Subscribe
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Subscribe Should handle jsonp error 2`] = `
<div
  class="subscribe"
>
  <form
    class="ui mini loading form"
    id="mailchimpForm"
  >
    <div
      class="unstackable fields"
    >
      <div
        class="required ten wide field"
      >
        <div
          class="ui input"
        >
          <input
            aria-label="email"
            id="email"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="admin@test.com"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <button
          class="ui mini secondary button"
        >
          Subscribe
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Subscribe Should subscribe to email 1`] = `
<div
  class="subscribe"
>
  <form
    class="ui mini loading form"
    id="mailchimpForm"
  >
    <div
      class="unstackable fields"
    >
      <div
        class="required ten wide field"
      >
        <div
          class="ui input"
        >
          <input
            aria-label="email"
            id="email"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="admin@test.com"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <button
          class="ui mini secondary button"
        >
          Subscribe
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Subscribe Should subscribe to email 2`] = `
<div
  class="subscribe"
>
  <form
    class="ui mini loading form"
    id="mailchimpForm"
  >
    <div
      class="unstackable fields"
    >
      <div
        class="required ten wide field"
      >
        <div
          class="ui input"
        >
          <input
            aria-label="email"
            id="email"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="admin@test.com"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <button
          class="ui mini secondary button"
        >
          Subscribe
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Subscribe Should update email with value 1`] = `
<div
  class="subscribe"
>
  <form
    class="ui mini form"
    id="mailchimpForm"
  >
    <div
      class="unstackable fields"
    >
      <div
        class="required ten wide field"
      >
        <div
          class="ui input"
        >
          <input
            aria-label="email"
            id="email"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="admin@test.com"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <button
          class="ui mini secondary button"
        >
          Subscribe
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Subscribe renders correctly no endpoint 1`] = `null`;

exports[`Subscribe renders correctly with endpoint 1`] = `
<div
  className="subscribe"
>
  <form
    className="ui mini form"
    id="mailchimpForm"
    onSubmit={[Function]}
  >
    <div
      className="unstackable fields"
    >
      <div
        className="required ten wide field"
      >
        <div
          className="ui input"
        >
          <input
            aria-label="email"
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            required={true}
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        className="field"
      >
        <button
          className="ui mini secondary button"
          onClick={[Function]}
        >
          Subscribe
        </button>
      </div>
    </div>
  </form>
</div>
`;
